File: React_Native/React_Native/src/pages/sat_pages/alterarCodigo.js
import React, { Component } from "react";
import { Text, View, Alert,Picker,TextInput,Button,DeviceEventEmitter } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

import OperacaoSat from '../../services/operacaoSat';
import RetornoSat from '../../services/retornoSat';

import ValidaCodigo from '../../components/ValidacaoCodigo';

export default class AlterarCodigo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            opcao:'Codigo de ativacao',
            codigoAtivacao:global.MyVar,
            codigoAtivacaoNovo:'',
            confirmarCodigo:'',
        };
    }

    dialogoSat(messageText){
        Alert.alert("Retorno", messageText);
    }
  
    // Função para validar os valores digitos pelo usuario e realizar a ativação do SAT
    alterarCodigoSat(){
        // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas 
        global.MyVar = this.state.codigoAtivacao

        // CRIA UMA INSTÂNCIA DO COMPONENTE QUE VALIDA O CÓDIGO
        var validationCodigo = new ValidaCodigo();

        if(validationCodigo.getValidation(this.state.codigoAtivacao) && validationCodigo.getValidation(this.state.codigoAtivacaoNovo)){
            if(this.state.codigoAtivacaoNovo == this.state.confirmarCodigo){
                // Caso todos os valores estejam corretos
                // Envia para o canal de ativação a função de alterar codigo do Sat
                var op = 0;

                // No Java o Sat reconhece 1 - Como sendo a opção de mudar codigo de ativação
                // 2 como sendo a opção de alterar codigo de emergência
                if(this.state.opcao == "Codigo de ativacao"){
                    op = 1;
                }else{
                    op = 2;
                }

                //ENVIA PARA OPERAÇÃO SAT ONDE VAI CONECTAR COM O SAT DE ACORDO COM A FUNÇÃP
                var operacaoSat = new OperacaoSat();
                args = {'funcao': 'TrocarCodAtivacao',
                        'random': Math.floor(Math.random() * 9999999).toString(),
                        'codigoAtivar': this.state.codigoAtivacao,
                        'codigoAtivarNovo': this.state.codigoAtivacaoNovo,
                        'op': op.toString()
                };
                operacaoSat.invocarOperacaoSat(args);

                DeviceEventEmitter.once("eventAlterar", event => { 
                    var retorno = event.alterar;
                    var retornoSat = new RetornoSat(args['funcao'], retorno);
                    this.dialogoSat(operacaoSat.formataRetornoSat(retornoSat));
                });
            }else{
                this.dialogoSat("O Código de Ativação Novo e a Confirmação do Código de Ativação não correspondem!");
            }
        } else{
            this.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!")
        }
    }

    render() {
        return (
            <SafeAreaView>
                <View style={{justifyContent:'space-between', flexDirection:'row', marginTop:20, marginBottom:10}}>
                    <Text style={{ fontSize:20, fontWeight:'bold' }} >Opções </Text>
                    <Picker 
                        selectedValue={this.state.opcao}
                        style={{height: 30, width: 200,marginEnd:100 }}
                        onValueChange={(itemValue, itemIndex) =>this.setState({opcao:itemValue})}
                    >      
                        <Picker.Item label="Código de ativação" value="Codigo de ativacao" />
                        <Picker.Item label="Código de emergencia" value="Codigo de emergencia" />
                    </Picker>
                </View>

                <View style={{ marginBottom:10,flexDirection:'row',justifyContent:'space-between' }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Atual</Text>
                    <TextInput
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:250, marginRight:50}}
                        value={this.state.codigoAtivacao}
                        onChangeText={(text) => this.setState({codigoAtivacao: text})}  
                    />  
                </View>

                <View style={{ marginBottom:10,flexDirection:'row',justifyContent:'space-between' }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Novo</Text>
                    <TextInput
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:250, marginRight:50}}
                        value={this.state.codigoAtivacaoNovo}
                        onChangeText={(text) => this.setState({codigoAtivacaoNovo: text})}  
                    />
                </View>

                <View style={{ marginBottom:10,flexDirection:'row',justifyContent:'space-between' }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Confirmar</Text>
                    <TextInput
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:250, marginRight:10}}            
                        value={this.state.confirmarCodigo}
                        onChangeText={(text) => this.setState({confirmarCodigo: text})}  
                    />
                </View>
                 
                <View style={{marginTop: 30, marginStart: 20, marginEnd: 20, marginBottom: 2}}>
                    <Button
                        color= 'gray'
                        title="alterar"
                        onPress={() => this.alterarCodigoSat()}
                    />
                </View>

            </SafeAreaView>   
        );
    }
}

File: Flutter/Flutter/lib/pages/sat_pages/alterarCodigo.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_gertec/services/globalValues.dart';
import 'package:flutter_gertec/services/operacaoSat.dart';
import 'package:flutter_gertec/services/retornoSat.dart';
import 'package:flutter_gertec/util/common_code.dart';
import 'package:flutter_gertec/widgets/widgetsgertec.dart';

class PageCodigoSat extends StatefulWidget {
  @override
  _PageCodigoSatState createState() => _PageCodigoSatState();
}

class _PageCodigoSatState extends State<PageCodigoSat> {
  final codigoAtivacaoNovo = TextEditingController(); // Codigo de ativação Novo

  // Inicializa o código de ativação com um valor global, para o usuario não precisar ficar digitando
  final codigoAtivacao = TextEditingController(text: GlobalValues.codAtivarSat); // Codigo de ativação Novo
  final confirmacaoCodigo = TextEditingController(); // Confirmação do codigo
  String tipoCodigo = "Código de ativação"; // Armazena o tipo de codigo

  CommonGertec commonGertec = new CommonGertec(); //* Classe que possui partes de código comuns em diversas telas

  // Função para validar os valores digitos pelo usuario e realizar a ativação do SAT
  void alterarCodigoSat() async {
    GlobalValues.codAtivarSat = codigoAtivacao.text; // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
    // Verifica se as entradas são validas
    if (commonGertec.isCodigoValido(codigoAtivacao.text) && commonGertec.isCodigoValido(codigoAtivacaoNovo.text)) {
      if (codigoAtivacaoNovo.text == confirmacaoCodigo.text) {
        // No Java o Sat reconhece 1 - Como sendo a opção de mudar codigo de ativação
        // 2 como sendo a opção de alterar codigo de emergência
        int op = 0;
        if (tipoCodigo == "Código de ativação")
          op = 1;
        else
          op = 2;

        //* Chama a função Invocar Operação Sat, que recebe como parâmetro a "operação invocada" e um "Map com as chaves e seus respectivos valores".
        RetornoSat retornoSat = await OperacaoSat.invocarOperacaoSat(
          // Passa como parâmetro um Map de argumentos de valores que deseja enviar
          args: {
            'funcao': "TrocarCodAtivacao",
            'random': Random().nextInt(999999), // Cada chamada no Java deve acompanhar diferente
            'codigoAtivar': codigoAtivacao.text.toString(),
            'codigoAtivacaoNovo': codigoAtivacaoNovo.text.toString(),
            'op': op
          },
        );
        //? Caso deseje visualizar o retorno completo basta chamar [retornoSat.getResultadoCompleto]

        //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
        //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
        // Recomenda-se acessar a função e entender como ela funciona
        String retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat: retornoSat);
        WidgetsGertec.dialogoSat(retornoFormatado, context: context);
      } else {
        WidgetsGertec.dialogoSat("O Código de Ativação Novo e a Confirmação do Código de Ativação não correspondem!", context: context);
      }
    } else {
      WidgetsGertec.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!", context: context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Container(
          color: Colors.white,
          padding: EdgeInsets.only(top: 30),
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              SizedBox(height: 30),
              dropDownList(),
              WidgetsGertec.formField(codigoAtivacao, "Atual: "),
              SizedBox(height: 20),
              WidgetsGertec.formField(codigoAtivacaoNovo, "Novo: "),
              SizedBox(height: 20),
              WidgetsGertec.formField(confirmacaoCodigo, "Confirmar: "),
              SizedBox(height: 20),
              SizedBox(
                width: 240,
                child: RaisedButton(
                  child: Text("ALTERAR"),
                  onPressed: () {
                    alterarCodigoSat();
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget dropDownList() {
    return Row(
      children: <Widget>[
        Text(
          "Tipo de codigo: ",
          style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
        ),
        DropdownButton<String>(
          value: tipoCodigo,
          icon: Icon(Icons.arrow_downward),
          iconSize: 24,
          elevation: 16,
          style: TextStyle(color: Colors.black, fontSize: 15),
          onChanged: (String newValue) {
            setState(
              () {
                tipoCodigo = newValue;
              },
            );
          },
          items: <String>[
            "Código de ativação",
            "Código de Emergência",
          ].map<DropdownMenuItem<String>>((String value) {
            return DropdownMenuItem<String>(
              value: value,
              child: Text(value.toString()),
            );
          }).toList(),
        ),
      ],
    );
  }
}
